mat/diagonal.go

```diff
@@ -85,41 +87,150 @@ func (d *DiagDense) TTri() Triangular {
 	return TransposeTri{d}
 }
 
+// TBand performs an implicit transpose by returning the receiver inside a
+// TransposeBand.
 func (d *DiagDense) TBand() Banded {
 	return TransposeBand{d}
 }
 
+// Bandwidth returns the upper and lower bandwidths of the matrix.
+// These values are always zero for diagonal matrices.
 func (d *DiagDense) Bandwidth() (kl, ku int) {
 	return 0, 0
 }
 
 // Symmetric implements the Symmetric interface.
 func (d *DiagDense) Symmetric() int {
-	return len(d.data)
+	return d.n
 }
 
 // Triangle implements the Triangular interface.
 func (d *DiagDense) Triangle() (int, TriKind) {
-	return len(d.data), Upper
+	return d.n, Upper
 }
 
+// Reset zeros the length of the matrix so that it can be reused as the
+// receiver of a dimensionally restricted operation.
+//
+// See the Reseter interface for more information.
+func (d *DiagDense) Reset() {
+	// No change of Inc or n to 0 may be
+	// made unless both are set to 0.
+	d.mat.Inc = 0
+	d.n = 0
+	d.mat.Data = d.mat.Data[:0]
+}
+
+// DiagOf extracts the diagonal of m into the receiver. If m can return
+// its raw representation d will be a view of the diagonal of m. Otherwise
+// the receiver will be a copy of the input's diagonal.
+// The receiver must be min(r, c) long or zero. Otherwise DiagOf will panic.
+func (d *DiagDense) DiagOf(m Matrix) {
+	// reuseAs is not used here since
+	// sensible cases should not allocate,
+	// but would uselessly if we did.
+	n := min(m.Dims())
+	if !d.IsZero() && n != d.n {
+		panic(ErrShape)
+	}
+	d.n = n
+
+	switch r := m.(type) {
+	case *DiagDense:
+		d.mat = r.mat
+	case RawBander:
+		mat := r.RawBand()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride,
+			Data: mat.Data[mat.KL : (n-1)*mat.Stride+mat.KL+1],
+		}
+	case RawMatrixer:
+		mat := r.RawMatrix()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride + 1,
+			Data: mat.Data[:(n-1)*mat.Stride+n],
+		}
+	case RawSymBander:
+		mat := r.RawSymBand()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride,
+			Data: mat.Data[:(n-1)*mat.Stride+1],
+		}
+	case RawSymmetricer:
+		mat := r.RawSymmetric()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride + 1,
+			Data: mat.Data[:(n-1)*mat.Stride+n],
+		}
+	case RawTriangular:
+		mat := r.RawTriangular()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride + 1,
+			Data: mat.Data[:(n-1)*mat.Stride+n],
+		}
+	case RawVectorer:
+		d.mat = r.RawVector()
+		d.mat.Data = d.mat.Data[:1]
+	default:
+		d.reuseAs(n)
+		for i := 0; i < n; i++ {
+			d.setDiag(i, m.At(i, i))
+		}
+	}
+}
+
+// RawBand returns the underlying data used by the receiver represented
```

Comment:
I know this function is already here, but is it a good idea? Currently in the addition types (or whatever), we switch on the type by which `Raw` it can implement. If methods implement multiple `RawBand` types, then we should make sure to order from the most specific to the least specific. Given that there's nothing wrong with implementing this method, I tend to lean toward fixing our ordering, but it's worth asking about.

Reply:
Agreed. We can have a note in the package level documentation about this.

I don't think most specific to least specific if the correct ordering though (Is there such an ordering in the general case? - I'm not sure there is). The ordering I think should exist is one that maintains correctness and gives the best performance available.